### PR TITLE
Add a warning to deepspeed when inferring batch size

### LINKED ISSUE #8872: Deepspeed accelerator calls datamodule.train_dataloader() prior to setup()
## 🐛 Bug

Deepspeed accelerator calls datamodule.train_dataloader() prior to setup(). This does not happen with other accelerators.

### To Reproduce

```py
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer, LightningDataModule


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class PlDataModule(LightningDataModule):
    def __init__(self):
        super().__init__()
        self._setup = False

    def setup(self, stage):
        self._setup = True

    def train_dataloader(self):
        assert self._setup
        return DataLoader(RandomDataset(32, 64), batch_size=2)

    def val_dataloader(self):
        assert self._setup
        return DataLoader(RandomDataset(32, 64), batch_size=2)

    def test_dataloader(self):
        assert self._setup
        return DataLoader(RandomDataset(32, 64), batch_size=2)


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        weights_summary=None,
        accelerator="deepspeed",
    )
    trainer.fit(model, datamodule=PlDataModule())


if __name__ == "__main__":
    run()
```

### Expected behavior

`train_dataloader` is never called before `setup`.

### Additional context

Backtrace

```
  /home/ubuntu/.local/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py(553)fit()
-> self._run(model)
  /home/ubuntu/.local/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py(865)_run()
-> self.accelerator.setup_environment()
  /home/ubuntu/.local/lib/python3.8/site-packages/pytorch_lightning/accelerators/gpu.py(30)setup_environment()
-> super().setup_environment()
  /home/ubuntu/.local/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py(76)setup_environment()
-> self.training_type_plugin.setup_environment()
  /home/ubuntu/.local/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp.py(166)setup_environment()
-> self.setup_distributed()
  /home/ubuntu/.local/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py(341)setup_distributed()
-> self._format_config()
  /home/ubuntu/.local/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py(545)_format_config()
-> self._format_batch_size_and_grad_accum_config()
  /home/ubuntu/.local/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py(555)_format_batch_size_and_grad_accum_config()
-> batch_size = self._auto_select_batch_size()
  /home/ubuntu/.local/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py(566)_auto_select_batch_size()
-> train_dataloader = self.lightning_module.train_dataloader()
```

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Closes #8872 

Adds a warning for users, as this may crash their code if they require setup to be called first.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 1]
```diff
@@ -562,6 +562,11 @@ def _format_batch_size_and_grad_accum_config(self):
                 " as this will be set via accumulate_grad_batches=x argument passed via the Lightning Trainer."
             )
         if "train_micro_batch_size_per_gpu" not in self.config:
+            rank_zero_warn(
+                "Inferring the batch size for internal deepspeed logging from the ``train_dataloader()``. "
+                "If you require skipping this, please pass "
+                "``Trainer(plugins=DeepSpeedPlugin(logging_batch_size_per_gpu=X)``, where X is the batch size."
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:568]: double backticks are for rendered docs

```suggestion
                "Inferring the batch size for internal deepspeed logging from the `train_dataloader()`. "
                "If you require skipping this, please pass "
                "`Trainer(plugins=DeepSpeedPlugin(logging_batch_size_per_gpu=batch_size)`"
```

**Review by carmocca — APPROVED** [review]

File: tests/plugins/test_deepspeed_plugin.py [hunk 2]
```diff
@@ -830,3 +830,55 @@ def test_deepspeed_multigpu_no_schedulers(tmpdir):
...
+        def train_dataloader(self):
+            assert self._setup
+            return DataLoader(RandomDataset(32, 64), batch_size=2)
+
+        def val_dataloader(self):
+            assert self._setup
+            return DataLoader(RandomDataset(32, 64), batch_size=2)
+
+        def test_dataloader(self):
+            assert self._setup
```
Comment (awaelchli) [inline tests/plugins/test_deepspeed_plugin.py:874]: this assert never gets called

File: tests/plugins/test_deepspeed_plugin.py [hunk 3]
```diff
@@ -830,3 +830,55 @@ def test_deepspeed_multigpu_no_schedulers(tmpdir):
...
+    """
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        plugins=[DeepSpeedPlugin()],
+        gpus=1,
+        fast_dev_run=True,
+    )
+    with pytest.warns(UserWarning, match="Inferring the batch size for internal deepspeed logging"):
+        trainer.fit(model)
```
Comment (awaelchli) [inline tests/plugins/test_deepspeed_plugin.py:848]: Isn't the test a bit expensive just for a warning?
We could probably unit test this directly by `calling setup_distributed()`. 
What do you think?

File: tests/plugins/test_deepspeed_plugin.py [hunk 4]
```diff
@@ -830,3 +830,55 @@ def test_deepspeed_multigpu_no_schedulers(tmpdir):
...
+        def __init__(self):
+            super().__init__()
+            self._setup = False
+
+        def setup(self, stage: Optional[str] = None) -> None:
+            self._setup = True
+
+        def train_dataloader(self):
+            assert self._setup
+            return DataLoader(RandomDataset(32, 64), batch_size=2)
```
Comment (awaelchli) [inline tests/plugins/test_deepspeed_plugin.py:867]: I think you want to test here that the dataloader() method gets called after setup when auto_select_batch_size gets used. 

But shouldn't that method better use the `trainer.train_dataloader`? Calling a LightningModule's dataloader method  several times may have unintended side effects.

Comment (SeanNaren) [reply tests/plugins/test_deepspeed_plugin.py:867]: I think the issue arises from the logic being tested here living in the pre_training routine, before the dataloader is attached to the trainer (within the fit loop advance).

Comment (awaelchli) [reply tests/plugins/test_deepspeed_plugin.py:867]: ah I see, the call is here https://github.com/PyTorchLightning/pytorch-lightning/blob/b046bd0670cd86300ad77533d300db79b137054a/pytorch_lightning/trainer/trainer.py#L1127 but that's too late. 

shame :(

File: tests/plugins/test_deepspeed_plugin.py [hunk 5]
```diff
@@ -830,3 +830,55 @@ def test_deepspeed_multigpu_no_schedulers(tmpdir):
...
+    Test DeepSpeed works when setup is required to call, and the user passes the batch size manually.
+    """
+
+    class PlDataModule(LightningDataModule):
+        def __init__(self):
+            super().__init__()
+            self._setup = False
+
+        def setup(self, stage: Optional[str] = None) -> None:
+            self._setup = True
```
Comment (tchaton) [inline tests/plugins/test_deepspeed_plugin.py:863]: It should be:

```suggestion
            self._setup = torch.distributed.is_available() and torch.distributed.is_initialized()
```

Comment (tchaton) [reply tests/plugins/test_deepspeed_plugin.py:863]: But this would fail for ddp_spawn. @awaelchli @ananthsub  I believe this is an important issue within Lightning right now.

File: tests/plugins/test_deepspeed_plugin.py (hunk 3 above)
Comment (SeanNaren) [reply tests/plugins/test_deepspeed_plugin.py:848]: I tried, however this was extremely difficult as references are required (lightning module primarily). Any hack arounds to make the unit test work I think will just make the test fragile however if you have suggestions let me know!

File: tests/plugins/test_deepspeed_plugin.py (hunk 5 above)
Comment (SeanNaren) [reply tests/plugins/test_deepspeed_plugin.py:863]: This doesn't apply to this module I think, this is only for DeepSpeed and Deepspeed doesnt support spawn!

File: tests/plugins/test_deepspeed_plugin.py [hunk 6]
```diff
@@ -830,3 +830,56 @@ def test_deepspeed_multigpu_no_schedulers(tmpdir):
     trainer.fit(model)
 
     _assert_save_model_is_equal(model, tmpdir, trainer)
+
+
+@RunIf(min_gpus=1, deepspeed=True, special=True)
+def test_deepspeed_warn_train_dataloader_called(tmpdir):
+    """
+    Test DeepSpeed warns when it calls ``train_dataloader`` internally for logging batch size.
+    """
```
Comment (awaelchli) [inline tests/plugins/test_deepspeed_plugin.py:839]: I think precommit complains about these docstrings

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 7]
```diff
@@ -546,6 +546,11 @@ def _format_batch_size_and_grad_accum_config(self):
                 " as this will be set via accumulate_grad_batches=x argument passed via the Lightning Trainer."
             )
         if "train_micro_batch_size_per_gpu" not in self.config:
+            rank_zero_warn(
+                "Inferring the batch size for internal deepspeed logging from the `train_dataloader()`. "
+                "If you require skipping this, please pass "
+                "`Trainer(plugins=DeepSpeedPlugin(logging_batch_size_per_gpu=batch_size)`"
```
Comment (ethanwharris) [inline pytorch_lightning/plugins/training_type/deepspeed.py:552]:
```suggestion
                "`Trainer(plugins=DeepSpeedPlugin(logging_batch_size_per_gpu=batch_size))`"
```

**Review by ethanwharris — APPROVED** [review]
LGTM, small comment :smiley:

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 7 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:552]: great catch!
